modules/cdp-engine/src/lib.rs

```diff
@@ -1215,59 +1269,207 @@ impl<T: Config> Pallet<T> {
+			.reciprocal()
+			.unwrap_or_else(Ratio::max_value)
+			.saturating_mul_int(
+				T::PriceSource::get_relative_price(T::GetStableCurrencyId::get(), currency_id)
+					.expect("the oracle price should be available because liquidation are triggered by it.")
+					.saturating_mul_int(target_stable_amount),
+			);
+		let collateral_supply = amount.min(max_supply_limit);
+
+		let collateral = currency_id
+			.erc20_address()
+			.ok_or(Error::<T>::CollateralContractNotFound)?;
+		let repay_dest = Pallet::<T>::evm_address();
+		let repay_dest_account_id = Pallet::<T>::account_id();
+
+		let stable_coin = T::GetStableCurrencyId::get();
+
+		let contracts_by_priority = {
+			let now: u32 = frame_system::Pallet::<T>::current_block_number()
+				.try_into()
+				.map_err(|_| ArithmeticError::Overflow)?;
+			// can't fail as ensured `liquidation_contracts_len` non-zero
+			let start_at = (now as usize) % liquidation_contracts_len;
+			let mut all: Vec<EvmAddress> = liquidation_contracts.into();
+			let mut right = all.split_off(start_at);
+			right.append(&mut all);
+			right
+		};
+
+		// try liquidation on each contract
+		for contract in contracts_by_priority.into_iter() {
+			let repay_dest_balance = CurrencyOf::<T>::free_balance(stable_coin, &repay_dest_account_id);
+			if T::LiquidationEvmBridge::liquidate(
+				InvokeContext {
+					contract,
+					sender: Pallet::<T>::evm_address(),
+					origin: contract,
+				},
+				collateral,
+				repay_dest,
+				collateral_supply,
+				target_stable_amount,
+			)
+			.is_ok()
+			{
+				let repayment = CurrencyOf::<T>::free_balance(stable_coin, &repay_dest_account_id)
+					.saturating_sub(repay_dest_balance);
+				let contract_account_id = T::EvmAddressMapping::get_account_id(&contract);
+				if repayment >= target_stable_amount {
+					// sufficient repayment, transfer collateral to contract and notify
+					if let Err(e) = <T as Config>::CDPTreasury::withdraw_collateral(
+						&contract_account_id,
+						currency_id,
+						collateral_supply,
+					) {
+						log::error!(
+							target: "cdp-engine",
+							"LiquidateViaContracts: transfer collateral to contract failed. \
+							Collateral: {:?}, amount: {:?} contract: {:?}, error: {:?}. \
+							This is unexpected, need extra action.",
+							currency_id, collateral_supply, contract, e,
+						);
+					}
+					// notify liquidation success
+					T::LiquidationEvmBridge::on_collateral_transfer(
+						InvokeContext {
+							contract,
+							sender: Pallet::<T>::evm_address(),
+							origin: contract,
+						},
+						collateral,
+						target_stable_amount,
+					);
+					// refund rest collateral to CDP owner
+					let refund_collateral_amount = amount
+						.checked_sub(collateral_supply)
+						.expect("Ensured collateral supply <= amount; qed");
+					if !refund_collateral_amount.is_zero() {
+						<T as Config>::CDPTreasury::withdraw_collateral(who, currency_id, refund_collateral_amount)?;
```

Comment:
I think if any of `method()?;` will fail it will return `Err` and not continue with other contracts

Reply:
updated